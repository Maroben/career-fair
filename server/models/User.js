const mongoose = require("mongoose")
const jwt = require("jsonwebtoken")
const config = require("config")
const Joi = require("@hapi/joi")

const userSchema = new mongoose.Schema({
	email: {
		type: String,
		required: true,
		unique: true,
		minlength: 3,
		maxlength: 255
	},
	password: {
		type: String,
		required: true,
		minlength: 5,
		maxlength: 1024
	},
	isAdmin: Boolean
})

const userJoi = {
	email: Joi.string()
		.min(6)
		.max(255)
		.required()
		.email(),
	password: Joi.string()
		.min(5)
		.max(255)
		.required()
}

userSchema.methods.generateAuthToken = () => {
	return jwt.sign(
		{ _id: this._id, email: this.email, isAdmin: this.isAdmin },
		config.get("jwtPrivateKey")
	)
}

module.exports = {
	Model: mongoose.model("users", userSchema),
	attr: [["email", "password"], ["_id", "email", "isAdmin"]],
	validate: (user) => {
		return Joi.validate(user, userJoi)
	}
}
