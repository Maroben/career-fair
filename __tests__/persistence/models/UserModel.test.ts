import _ from "lodash"
import User, { IUser, validate, properties, level } from "../../../src/persistence/models/UserModel"

describe("User Model", () => {
    describe("Valid User Creations", () => {
        it("Simple User", () => {
            const user: IUser = new User({
                email: "maroben@example.com",
                password: "security_lv_9000",
                level: level.user
            })
            expect(_.pick(user, properties)).toMatchSnapshot()
            expect(async () => await validate(user)).resolves
        })
    })

    describe("Joi User Validation Rejections", () => {
        it("Empty Object", () => {
            const user: IUser = new User({})
            expect(async () => await validate(user)).rejects
        })

        describe("Email", () => {
            it("Empty", () => {
                const user: IUser = new User({
                    email: "",
                    password: "very secure",
                    level: level.user
                })
                expect(async () => await validate(user)).rejects
            })

            it("wrong format 1", () => {
                const user: IUser = new User({
                    email: "this is not an email",
                    password: "very secure",
                    level: level.admin
                })
                expect(async () => await validate(user)).rejects
            })

            it("wrong format 2", () => {
                const user: IUser = new User({
                    email: "fake@totoally",
                    password: "very secure",
                    level: level.root
                })
                expect(async () => await validate(user)).rejects
            })

            it("wrong format 3", () => {
                const user: IUser = new User({
                    email: "might@look,correct",
                    password: "very secure",
                    level: level.user
                })
                expect(async () => await validate(user)).rejects
            })

            it("wrong format 4", () => {
                const user: IUser = new User({
                    email: "might(at)look.correct",
                    password: "very secure",
                    level: level.user
                })
                expect(async () => await validate(user)).rejects
            })
        })

        describe("Password", () => {
            it("Empty", () => {
                const user: IUser = new User({
                    email: "maroben@example.com",
                    password: "",
                    level: level.user
                })
                expect(async () => await validate(user)).rejects
            })

            it("Too long", () => {
                const user: IUser = new User({
                    email: "maroben@example.com",
                    password: "wayyyyy tooooo looooong",
                    level: level.user
                })
                expect(async () => await validate(user)).rejects
            })
        })

        describe("Permission Level", () => {
            it("Admin", () => {
                const user: IUser = new User({
                    email: "maroben@example.com",
                    password: "",
                    level: level.admin
                })
                expect(async () => await validate(user)).toMatchSnapshot()
            })

            it("negative", () => {
                const user: IUser = new User({
                    email: "maroben@example.com",
                    password: "",
                    level: level.root - 1
                })
                expect(async () => await validate(user)).rejects
            })

            it("too high", () => {
                const user: IUser = new User({
                    email: "maroben@example.com",
                    password: "",
                    level: level.user + 1
                })
                expect(async () => await validate(user)).rejects
            })
        })

        it("Wrong data type", () => {
            const user: IUser = new User({
                email: 8,
                password: {},
                level: "admin"
            })
            expect(async () => await validate(user)).rejects
        })
    })
})
