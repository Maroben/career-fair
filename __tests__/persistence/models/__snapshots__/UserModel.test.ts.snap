// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Model Joi User Validation Rejections Permission Level Admin 1`] = `[Function]`;

exports[`User Model Valid User Creations Simple User 1`] = `
Object {
  "email": "maroben@example.com",
  "level": 2,
  "password": "security_lv_9000",
}
`;
